Cumulus 1482 data model changes (#1379)
nasa/cumulus: DATA_README.md

# Cumulus - Data model

Cumulus uses DynamoDB for database storage. While DynamoDB is a schemaless database storage system, Cumulus enforces schemas on record creation and update using the [`ajv` package](https://github.com/epoberezkin/ajv).

The data models managed by Cumulus and their schema definitions can be found in the [Cumulus source code](./packages/api/models/schemas.js).

## Managing data schemas

By default, all changes to Cumulus data models should be **backwards-compatible with all previous versions of the data models since the last data migration (if any)**. By "backwards-compatible", we mean:

> - Cumulus API endpoint responses have the same structure and properties

Author: What else should be listed here?

Thus, any changes to the underlying data model schemas would require corresponding functions to translate between the updated schemas and all previous versions of the schemas so that data read from the models would conform to the same format.

> Backwards compatibility is required because all data that Cumulus manages must be readable and usable in a consistent format. Furthermore, given that Cumulus is actively being used in operational environments which are continuously ingesting petabytes of data, if backwards compatibility were not enforced it would not be realistic to assume that all previously ingested data would be re-ingested to conform to the updated schemas.

Author: I'm struggling to determine for myself **why** Cumulus data must be output in a consistent format. Why can't our API endpoint response change?

Reviewer: I think we could change the API response, but we want to minimize those changes to minimize the impact to those who may have coded around it. If we had an API that was just out there we'd have to be  backwards compatible or go to v2 of the API. But since everyone manages their own deployments we could release changes in a "major" version release, since it would be breaking. Originally (2 years ago) we were talking about API versioning, but I'm not sure that is so necessary.

The concern for this ticket is if we changed something in `api/models/` and workflows were in flight and they were expecting say collection to be in a certain structure and the structure changed and the workflow was using lambdas that had not been updated.

> # Cumulus - Data model

Reviewer 2: Same vis-a-vis naming, I'd prefer `DATA_MODEL_README.md` or something otherwise including `DATA_MODEL` vs. just `DATA`.

Cumulus uses DynamoDB for database storage. While DynamoDB is a schemaless database storage system, Cumulus enforces schemas on record creation and update using the [`ajv` package](https://github.com/epoberezkin/ajv).

The data models managed by Cumulus and their schema definitions can be found in the [Cumulus source code](./packages/api/models/schemas.js).

## Documenting changes

> All changes to the database models should be noted in the [data model CHANGELOG](./DATA_CHANGELOG.md).

Reviewer 2: ```suggestion
All changes to the database models must be noted in the [data model CHANGELOG](./DATA_CHANGELOG.md). Tracking backwards compatibility will otherwise be extremely difficult.
```

## Managing data schemas

By default, all changes to Cumulus data models should be **backwards-compatible with all previous versions of the data models since the last data migration (if any)**.

By "backwards-compatible", we mean that **any new required fields must be nullable or translatable**.

Making new field(s) translatable means that they must be able to be populated from data contained in **any previous version of the database schema**.

> Enforcing backwards compatibility means that database operations can conform to the following rules:

Reviewer 2: ```suggestion
Enforcing backwards compatibility means that database operations must conform to the following rules:
```

However, there would be other harmful consequences of deploying breaking schema changes:

- [Workflow reporting](./docs/interfaces.md) Lambdas, which are not versioned, could receive data from a long-running workflow in an old schema format but attempt to save the record according to the new schema, which would fail. Thus, even "successful" workflows would not have any archived records associated with them nor be visible to Cumulus operators
- API responses would be inconsistent, which could be problematic for API consumers

While it is technically possible to version API endpoints to isolate different responses and to prevent impacting API consumers, producing a consistent response even for a given API version would require a migration script that would convert data from all schema versions prior to the breaking change to conform to the latest schema. Otherwise, even requests to a versioned API endpoint could produce inconsistent responses from the prior schema versions.

> 

Reviewer 2: 👍 Great section!

### Supporting breaking changes

> If for some reason, schema changes must be made that cannot be made backwards compatible, then this change needs to be discussed in an architecture meeting of the Cumulus core development team.

Reviewer 2: ```suggestion
Any non-backwards compatible schema changes that cannot be avoided must be discussed and approved in an architecture meeting of the Cumulus core development team.
```
